src/parser.rs

```diff
@@ -423,6 +417,61 @@ impl Parser {
         })
     }
 
+    /// Parse a SQL LISTAGG expression, e.g. `LISTAGG(...) WITHIN GROUP (ORDER BY ...)`.
+    pub fn parse_listagg_expr(&mut self) -> Result<Expr, ParserError> {
+        self.expect_token(&Token::LParen)?;
+        let distinct = self.parse_all_or_distinct()?;
+        let expr = Box::new(self.parse_expr()?);
+        // While ANSI SQL would would require the separator, Redshift makes this optional. Here we
+        // choose to make the separator optional as this provides the more general implementation.
+        let separator = if self.consume_token(&Token::Comma) {
+            Some(Box::new(self.parse_expr()?))
+        } else {
+            None
+        };
+        let on_overflow = if self.parse_keywords(vec!["ON", "OVERFLOW"]) {
+            if self.parse_keyword("ERROR") {
+                Some(ListAggOnOverflow::Error)
+            } else {
+                self.expect_keyword("TRUNCATE")?;
+                let filler = match self.peek_token() {
+                    Some(Token::Word(kw)) if kw.keyword == "WITH" || kw.keyword == "WITHOUT" => {
+                        None
+                    }
+                    Some(Token::SingleQuotedString(_)) => Some(Box::new(self.parse_expr()?)),
+                    _ => self.expected("either filler, WITH, or WITHOUT", self.peek_token())?,
+                };
+                let with_count = self.parse_keyword("WITH");
+                if !with_count {
+                    self.expect_keyword("WITHOUT")?;
+                }
```

Comment:
I reworked this a bit since now the check for `WITH` or `WITHOUT` should be happening on line 442.

Reply:
The WITH/WITHOUT lookahead you're referring to should not affect this code, as this can execute after the lookahead failed and we've parsed the 'filler'.

This code is equivalent to the previous one in the happy case, but will report a slightly confusing error of "Expected keyword WITHOUT, found FOO"  when parsing `'filler' FOO`.

Reply:
Ah shoot, good catch! I'm not sure why I keep running into this.